File: proedc/Java/proedc.java
package com.brainai.proedc;

import com.sun.jna.*;
import com.sun.jna.ptr.PointerByReference;
import java.util.Arrays;
import java.util.List;

public class ProNative {

    // --- ENUMS & KONSTANTEN ---
    public interface ProED_Mode {
        int ENCRYPT = 0;
        int DECRYPT = 1;
    }

    public interface ProEDC_Status {
        int SUCCESS = 0;
        int ERR_PARAM = -1;
        int ERR_FILE_IO = -2;
        int ERR_CRYPTO = -3;
        int ERR_HASH = -4;
        int ERR_KEYGEN = -5;
        int ERR_MEMORY = -6;
        int ERR_LIMIT_REACHED = -7;
    }

    // --- STRUKTUREN (C-Kompatibel) ---

    @Structure.FieldOrder({"state", "stream", "idx", "mode"})
    public static class ProED_Context extends Structure {
        public int[] state = new int[8];      // uint32_t state[8]
        public byte[] stream = new byte[32];  // uint8_t stream[32]
        public int idx;                       // uint32_t idx
        public int mode;                      // uint32_t mode (Enum)

        public ProED_Context() {
            super();
            setAlignType(ALIGN_GNUC); // 4-Byte Alignment
        }
    }

    @Structure.FieldOrder({"belt", "count", "buffer", "bufIdx"})
    public static class ProHash_Ctx extends Structure {
        public int[] belt = new int[16];      // uint32_t belt[16]
        public long count;                    // uint64_t count
        public byte[] buffer = new byte[64];  // uint8_t buffer[64]
        public int bufIdx;                    // uint32_t buf_idx

        public ProHash_Ctx() {
            super();
            setAlignType(ALIGN_DEFAULT); // 8-Byte Alignment für uint64_t
        }
    }

    // --- INTERFACE DEFINITIONEN ---

    public interface IProKey extends Library {
        IProKey INSTANCE = Native.load("prokey", IProKey.class);
        void ProKey_Generate(byte[] buffer, NativeSize length);
    }

    public interface IProHash extends Library {
        IProHash INSTANCE = Native.load("prohash", IProHash.class);
        void prohash_init(ProHash_Ctx ctx);
        void prohash_update(ProHash_Ctx ctx, byte[] data, NativeSize len);
        void prohash_final(ProHash_Ctx ctx, byte[] digest);
    }

    public interface IProED extends Library {
        IProED INSTANCE = Native.load("proed", IProED.class);
        void proed_init(ProED_Context ctx, byte[] key, NativeSize keyLen, byte[] iv, NativeSize ivLen, int mode);
        void proed_process(ProED_Context ctx, byte[] data, NativeSize len);
    }

    public interface IProEDC extends Library {
        IProEDC INSTANCE = Native.load("proedc", IProEDC.class);
        void ProEDC_Init();
        int ProEDC_EncryptFile(String inputFile, String outputFile, String keyFile);
        int ProEDC_DecryptFile(String inputFile, String outputFile, String keyFile);
        String ProEDC_GetVersion();
    }

    // --- HIGH LEVEL HELPER ---
    public static class Client {
        static {
            IProEDC.INSTANCE.ProEDC_Init();
        }

        public static byte[] generateKey(int bytes) {
            byte[] key = new byte[bytes];
            IProKey.INSTANCE.ProKey_Generate(key, new NativeSize(bytes));
            return key;
        }

        public static String calculateHash(byte[] data) {
            ProHash_Ctx ctx = new ProHash_Ctx();
            IProHash.INSTANCE.prohash_init(ctx);
            IProHash.INSTANCE.prohash_update(ctx, data, new NativeSize(data.length));
            byte[] digest = new byte[32];
            IProHash.INSTANCE.prohash_final(ctx, digest);
            
            StringBuilder sb = new StringBuilder();
            for (byte b : digest) sb.append(String.format("%02X", b));
            return sb.toString();
        }

        public static int encryptFile(String in, String out, String key) {
            return IProEDC.INSTANCE.ProEDC_EncryptFile(in, out, key);
        }
    }
}

File: proedc/JavaScript/proedc.js
const ffi = require('ffi-napi');
const ref = require('ref-napi');
const StructType = require('ref-struct-napi');
const ArrayType = require('ref-array-napi');

// --- TYP-DEFINITIONEN ---
const uint32 = ref.types.uint32;
const uint8 = ref.types.uint8;
const uint64 = ref.types.uint64;
const size_t = ref.types.size_t; // Plattformabhängig (4 oder 8 Byte)

// Arrays für die Structs
const Uint32Array8 = ArrayType(uint32, 8);
const Uint32Array16 = ArrayType(uint32, 16);
const Uint8Array32 = ArrayType(uint8, 32);
const Uint8Array64 = ArrayType(uint8, 64);

// --- STRUCTS ---

// Entspricht ProED_Context (aus ProED_Interface.h)
const ProEDContext = StructType({
    'state': Uint32Array8,
    'stream': Uint8Array32,
    'idx': uint32,
    'mode': uint32
});

// Entspricht ProHash_Ctx
const ProHashCtx = StructType({
    'belt': Uint32Array16,
    'count': uint64,
    'buffer': Uint8Array64,
    'buf_idx': uint32
});

// --- LIBRARY LOADING ---
// Erkennt automatisch die Endung (.dll, .so) basierend auf dem OS
const libPath = (name) => process.platform === 'win32' ? name : `./lib${name}.so`;

const prokey = ffi.Library(libPath('prokey'), {
    'ProKey_Generate': ['void', ['pointer', size_t]]
});

const prohash = ffi.Library(libPath('prohash'), {
    'prohash_init': ['void', [ref.refType(ProHashCtx)]],
    'prohash_update': ['void', [ref.refType(ProHashCtx), 'pointer', size_t]],
    'prohash_final': ['void', [ref.refType(ProHashCtx), 'pointer']]
});

const proed = ffi.Library(libPath('proed'), {
    'proed_init': ['void', [ref.refType(ProEDContext), 'pointer', size_t, 'pointer', size_t, uint32]],
    'proed_process': ['void', [ref.refType(ProEDContext), 'pointer', size_t]]
});

const proedc = ffi.Library(libPath('proedc'), {
    'ProEDC_Init': ['void', []],
    'ProEDC_EncryptFile': ['int', ['string', 'string', 'string']],
    'ProEDC_DecryptFile': ['int', ['string', 'string', 'string']],
    'ProEDC_GetVersion': ['string', []]
});

// --- HIGH LEVEL API ---

class ProEDC {
    constructor() {
        proedc.ProEDC_Init();
    }

    getVersion() {
        return proedc.ProEDC_GetVersion();
    }

    /**
     * Erzeugt einen sicheren Zufallsschlüssel
     */
    generateKey(length = 32) {
        const buffer = Buffer.alloc(length);
        prokey.ProKey_Generate(buffer, length);
        return buffer;
    }

    /**
     * Berechnet den ProHash-256 Digest einer Datenmenge
     */
    calculateHash(data) {
        const ctx = new ProHashCtx();
        prohash.prohash_init(ctx.ref());

        const input = Buffer.isBuffer(data) ? data : Buffer.from(data);
        prohash.prohash_update(ctx.ref(), input, input.length);

        const digest = Buffer.alloc(32);
        prohash.prohash_final(ctx.ref(), digest);
        return digest.toString('hex').toUpperCase();
    }

    /**
     * Verschlüsselt eine Datei (Status-Codes gemäß ProEDC.h)
     */
    encryptFile(inputFile, outputFile, keyFile) {
        return proedc.ProEDC_EncryptFile(inputFile, outputFile, keyFile);
    }
}

module.exports = new ProEDC();
